053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+# Summary
+
+This RFC proposes a way to support pluggable external stores for build events
+and to provide a general mechanism for processing build events.
+
+
+# Motivation
+
+The first (and primary) motivation for this change is to allow operators to
+configure external data store for build events. They may want to do this for a
+number of reasons:
+
+1. Builds can generate an enormous amount of build events - storing everything
+   in the same database as all of the other core-Concourse information means
+   that, over-time, the `build_events` tables will tend to dominate (in many
+   cases). For instance, in our (fairly small) Concourse deployment at
+   `ci.concourse-ci.org`, build events comprise ~70% of the total ~13GB used.
+   While storage is cheap, I can imagine this could result in general sluggishness
+   in larger Concourse deployments as indexes get large. While much of this can be
+   alleviated by more aggressive build log reaping, that can introduce performance
+   problems of its own, since deleting lots of build events can be taxing on
+   Postgres. I'm sure it's fairly obvious at this point that I really don't know
```

Comment:
Both the performance overhead and deletion cost would be improved by using table partitions.

Reply:
I thought that was a fairly new feature, but it seems like it isn't:

<img width="1408" alt="Screen Shot 2020-05-12 at 2 09 31 PM" src="https://user-images.githubusercontent.com/3574444/81745784-42636880-945a-11ea-92fb-a4409a35e9e8.png">

that's nice (saying that especially because https://github.com/concourse/concourse/issues/5391#issuecomment-607346689)

Reply:
Yes, though the advances in table partition capabilities and performance between each version are substantial.

(Also the Concourse postgresql support policy should change, but that's another discussion).

Reply:
In particular:

1. Foreign key and uniqueness constraints were not enforced across partitions until v11.
1. Multiple partition keys were not supported until I think 11 or 12. These let you partition the main multiple ways if it proves to be useful. Partitioning by team and date seems reasonable, but studying existing production databases would be better.

Reply:
in the Partitioning doc for 9.5/9.6, it seems to be what we're doing (creating a table for each pipeline/team that `INHERITS` from the `build_events` table)

@jchesterpivotal could you elaborate on what you mean by table partitions? I notice in Postgres 10+ there's "declarative partitioning" (e.g. `CREATE TABLE ... PARTITION BY ...`) - is that what you're referring to?

Reply:
"Table partitioning" is the general name for the feature set of "some way of physically subdividing tables while giving some seamless single logical view". You can do it with table inheritance but it wasn't really designed for that purpose. The later features are more "real".

Especially because declarative partitioning is handled explicitly by the database, which can continuously find new performance gains in each release. With inheritance+triggers insert performance will be bound by your own code. And similarly at query time: a declarative partitioning scheme gives the query planner reliable information about which partitions can be ignored in a query plan.

Reply:
Though, I guess with our current Postgres support policy, we'll be stuck with table inheritance until 9.6 is no longer supported (not until late 2021) 🤷